File: c/src/operations/quick_sort.c
/**
 * Implementation of QuickSort.
 */

#include "quick_sort.h"

// Function to swap two elements
inline void swap(char** a, char** b) {
    char* temp = *a;
    *a = *b;
    *b = temp;
}

// Partition function to find the pivot position
inline int partition(char *arr[], int len) {
    char *pivot = arr[len - 1]; // Choose the last element as pivot
    int store_index = 0;

    for (int i = 0; i < len - 1; i++) {
        if (strcmp(arr[i], pivot) < 0) {    // Compare strings
            swap(&arr[i], &arr[store_index]);
            store_index++;
        }
    }
    swap(&arr[store_index], &arr[len - 1]); // Move pivot to its final place
    return store_index;
}

char** quick_sort(char *arr[], int len) {
    if (len > 1) {
        int pivot = partition(arr, len);
        quick_sort(arr, pivot);             // Recursively sort the left part
        quick_sort(arr + pivot + 1, len - pivot - 1); // Recursively sort the right part
    }
    return arr;
}

File: languages/java/app/src/main/java/org/benchmarkplayground/operations/QuickSort.java
package org.benchmarkplayground.operations;

import java.util.List;

public final class QuickSort {
    private static <T extends Comparable<T>> void swap(List<T> arr, int i, int j) {
        T temp = arr.get(i);
        arr.set(i, arr.get(j));
        arr.set(j, temp);
    }

    private static <T extends Comparable<T>> int partition(List<T> arr, int low, int high) {
        int pivot = high;   // Last element is pivot
        int i = low - 1;    // Index of smaller element

        for (int j = low; j < high; j++) {
            if (arr.get(j).compareTo(arr.get(pivot)) < 0) {
                i++;
                swap(arr, i, j);
            }
        }

        i++;
        swap(arr, i, pivot);
        return i;
    }

    private static <T extends Comparable<T>> List<T> quickSort(List<T> arr, int low, int high) {
        if (low < high) {
            int pivotIndex = partition(arr, low, high);
            quickSort(arr, low, pivotIndex - 1);
            quickSort(arr, pivotIndex + 1, high);
        }
        return arr;
    }

    public static <T extends Comparable<T>> List<T> quickSort(List<T> arr) {
        return quickSort(arr, 0, arr.size() - 1);
    }
}
